posts/searching-for-theorems-in-mathlib.md

```diff
@@ -0,0 +1,112 @@
+---
+author: 'Bolton Bailey'
+category: 'Overview'
+date: 2025-05-15 1:26:31 UTC-08:00
+description: ''
+has_math: false
+link: ''
+slug: searching-for-theorems-in-mathlib
+tags: ''
+title: Searching for Theorems in Mathlib
+type: text
+---
+A post for beginners on the different ways to search for theorems in mathlib, inspired by [this talk](https://www.youtube.com/watch?v=UJrYKR01QwU)  from Jireh Loreaux.
+
+<!-- TEASER_END -->
+
+Mathlib has a lot of theorems, and it can be hard to find the one you're looking for. This post goes over various search engines, webpages and tactics you can use to find theorems in mathlib. For each of them, I'll give a brief description of how they work and what their strengths and weaknesses are.
+
+## Search engines
+
+There are three main search engines which allow you to search for theorems in mathlib and its dependencies.
+
+### The mathlib4 docs page
+
+![Docs Main Page](/images/documentation-screenshot.png)
+
+The most comprehensive listing of mathlib theorems is the [documentation page on the community website](https://leanprover-community.github.io/mathlib4_docs/). This page was developed for Lean 3 by Rob Lewis and for Lean 4 by Henrik Böving and others, see the [docGen source for a list of contributors](https://github.com/leanprover/doc-gen4). It's a documentation page just as you might find for any other software project: on the left, it has nested dropdown menus for the directory structure of mathlib and several of its dependencies. In the upper right, it has a search bar. The search bar provides a "fuzzy" search which searches for theorem names that contain the search term as a substring. For example, if I search for ["**foobar**"](https://leanprover-community.github.io/mathlib4_docs/search.html?sitesearch=https%3A%2F%2Fleanprover-community.github.io%2Fmathlib4_docs&q=foobar), the current top result is [**F**GM**o**duleCat.**ob**j_c**ar**rier](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Category/FGModuleCat/Basic.html#FGModuleCat.obj_carrier).
+
+The docs are most useful for users who have some experience in guessing the name of the lemma they are looking for. Helpful in learning how to do this is the [naming conventions page](https://leanprover-community.github.io/contribute/naming.html) which explains how lemmas in mathlib are named (in short, theorems are typically named for the declarations they pertain to, with the declaration names appearing in the lemma name in the order they appear in the syntax tree).
+
+A downside of the documentation page is that if a lemma is named in a way you don't expect, it can be hard to find.
+
```

Comment:
I would mention tab completion somewhere, either here because it's the same kind of fuzzy algorithm as in the docs page, or when you talk about tactics because it's something you do while doing some Lean.